Use symlink xattr functions from x/sys/unix

tklauser opened the pull request:
Update x/sys/unix to get the symlink xattr syscall wrappers Lgetxattr and Lsetxattr (introduced in
golang/sys@b90f89a) and switch moby to use these instead of providing own syscall wrappers.

tophj-ibm commented:
failure looks related. I think the changes in golang.org/x/sys/windows/ztypes_windows.go are breaking it.
```
07:47:01 # github.com/docker/docker/pkg/term
07:47:01 pkg/term/term_windows.go:82: constant 4294967286 overflows int
07:47:01 pkg/term/term_windows.go:88: constant 4294967285 overflows int
07:47:01 pkg/term/term_windows.go:94: constant 4294967284 overflows int
07:47:12 Build step 'Execute shell' marked build as failure
```

tophj-ibm commented:
found the cl that changed it, might be useful https://go-review.googlesource.com/#/c/39608/

tklauser commented on pkg/system/xattrs_linux.go:
```diff
@@ -1,35 +1,21 @@
 // Lgetxattr retrieves the value of the extended attribute identified by attr
 // and associated with the given path in the file system.
 // It will returns a nil slice and nil error if the xattr is not set.
 func Lgetxattr(path string, attr string) ([]byte, error) {
```
Ok, will fold it in.

tklauser replied:
@AkihiroSuda I guess you meant to remove the Lsetxattr wrapper function? The Lgetxattr wrapper has additional functionality (resizing the buffer if the xattrs don't fit) which is not covered by the version from x/sys/unix.

AkihiroSuda replied:
ok, makes sense

tklauser replied:
Just checked. The wrapper is still necessary, as there is also a version in pkg/system/xattrs_unsupported.go for !linux.

tklauser commented:
@tophj-ibm Thanks for the hint. I currently see two possible solutions:

1. Switch GetStdFile in github.com/Azure/go-ansiterm/winterm to take an uint32 instead of int and check against windows.STD_*_HANDLE from golang.org/x/sys/windows.
2. (Temporarily) replace windows.STD_*_HANDLE in pkg/term/term_windows with either the values from syscall or define them locally as ints.

What would be the preferred way to resolve this? Would it be acceptable to implement option 2 as part of this PR and implement 1, once a fix is merged in github.com/Azure/go-ansiterm/winterm. Or does anyone see another option?

tophj-ibm commented:
@tklauser I am not a maintainer (but @AkihiroSuda is), but IMO option 2 where you switch the file to use syscall instead of golang.org/x/sys/windows and then open an issue for the first problem that seems good to me. 

I don't know if they will do anything about it in the go-ansiterm repo because they would have to switch their entire project from syscall to /x/sys/ in order to fix it, so the temporary solution might be sticking around for a while.

tklauser committed:
vendor: re-vendor golang.org/x/sys

Update golang.org/x/sys to 739734461d1c916b6c72a63d7efda2b27edb369f in
order to get the Lgetxattr, Llistxattr, Lremovexattr and Lsetxattr
syscall wrappers. These will be used in a successive commit to replace
the wrappers in libcontainer/system/xattr_linux.go.

Signed-off-by: Tobias Klauser <tklauser@distanz.ch>

tklauser committed:
[pkg/term] temporarily use STD_*_HANDLE from syscall again

Due to the CL https://go-review.googlesource.com/c/39608/ in
x/sys/windows which changed the definitions of STD_INPUT_HANDLE,
STD_OUTPUT_HANDLE and STD_ERROR_HANDLE, we get the following failure
after re-vendoring x/sys/windows:

  07:47:01 # github.com/docker/docker/pkg/term
  07:47:01 pkg/term/term_windows.go:82: constant 4294967286 overflows int
  07:47:01 pkg/term/term_windows.go:88: constant 4294967285 overflows int
  07:47:01 pkg/term/term_windows.go:94: constant 4294967284 overflows int
  07:47:12 Build step 'Execute shell' marked build as failure

Temporarily switch back pkg/term to use these constants from the syscall
package and add a comment about it.

To really fix this, go-ansiterm should probably be switched to use
x/sys/windows.

Signed-off-by: Tobias Klauser <tklauser@distanz.ch>

tklauser committed:
system: Use symlink xattr functions from x/sys/unix

Use the symlink xattr syscall wrappers Lgetxattr and Lsetxattr from
x/sys/unix (introduced in golang/sys@b90f89a) instead of providing own
wrappers. Leave the functionality of system.Lgetxattr intact with
respect to the retry with a larger buffer, but switch it to use
unix.Lgetxattr. Also leave system.Lsetxattr intact (even though it's
just a wrapper around the corresponding function from unix) in order to
keep moby building for !linux.

Signed-off-by: Tobias Klauser <tklauser@distanz.ch>

tklauser commented:
@AkihiroSuda Just pushed another version with an additional commit implementing option 2 outlined in https://github.com/moby/moby/pull/33539#issuecomment-306709227 (temporarily switching to use STD_*_HANDLE from syscall) as @tophj-ibm suggested. All builds looks fine now.

tophj-ibm commented:
changes SGTM

thaJeztah commented on pkg/term/term_windows.go:
```diff
@@ -78,20 +78,23 @@ func StdStreams() (stdIn io.ReadCloser, stdOut, stdErr io.Writer) {
 		emulateStderr = false
 	}
 
+	// Temporarily use STD_INPUT_HANDLE, STD_OUTPUT_HANDLE and
```
Should we mark this with a `TODO` ?

tklauser replied:
Makes sense, will add a `TODO`.

thaJeztah commented:
@tklauser looks like this needs a rebase 😢

tklauser commented:
@thaJeztah, rebased upon latest master and added the TODO as mentioned in https://github.com/moby/moby/pull/33539#discussion_r121666531

tklauser commented:
Rebased upon latest master again to resolve merge conflicts

thaJeztah commented:
ping @johnstep PTAL

sadly needs another rebase @tklauser  - I'll try to get someone review this ASAP, otherwise ping me 👍

tklauser commented:
Rebased again, but sadly the Windows build now fails due to the change in x/sys/windows introduced by  golang/sys@ff24cb3cd86e67842e0d078a5e0cb92d9ffcf8b4 (https://go-review.googlesource.com/c/39608/)

Any advise on how to properly fix this?

tklauser committed:
Windows: fix build after re-vendoring golang.org/x/sys

Due to the CL https://go-review.googlesource.com/c/39608/ in
x/sys/windows which changed the definitions of STD_INPUT_HANDLE,
STD_OUTPUT_HANDLE and STD_ERROR_HANDLE, we get the following failure
in cmd/dockerd/service_windows.go after re-vendoring x/sys/windows:

  06:29:57 # github.com/docker/docker/cmd/dockerd
  06:29:57 .\service_windows.go:400: cannot use sh (type int) as type uint32 in argument to windows.GetStdHandle

Fix it by adding an explicit type conversion when calling
windows.GetStdHandle.

Signed-off-by: Tobias Klauser <tklauser@distanz.ch>

tklauser commented:
Just pushed an additional commit adding an explicit type conversion in the call to `windows.GetStdHandle`. Hope that is OK...

tophj-ibm commented:
changes still seem fine to me.

thaJeztah commented:
For a list of commits in the golang.org/x/sys bump; https://github.com/golang/sys/compare/8f0908ab3b2457e2e15403d3697c9ef5cb4b57a9...739734461d1c916b6c72a63d7efda2b27edb369f

thaJeztah reviewed (approved):
LGTM

dnephin reviewed (approved):
LGTM

thaJeztah commented:
all green - merging. Thanks @tklauser !
